### DISCUSSION TITLE
Restricting HTTP METHODS via routes doesn't work per docs

### DISCUSSION THREAD

**Discussion by ctrlplusb** [vercel/vercel#4487] [Help] [answered]
[Docs state](https://vercel.com/docs/configuration#routes/methods) that the following would restrict the `/api/user.js` function so that it would only receive GET or POST requests.

```json
{
  "routes": [
    {
      "src": "/api/user.js",
      "methods": ["POST", "GET"],
      "dest": "/api/user.js"
    }
  ]
}
```

I did a basic deployment with a function that output the method like so:

```javascript
module.exports = (req, res) => {
  res.send(req.method);
};
```

Upon testing via [Insomnia](https://insomnia.rest/) I was able to hit the endpoint with any HTTP method.

**Comment by styfle** [comment, marked as answer]
Hi @ctrlplusb 

The docs don't specially mention restricting here but I see how this is confusing.

The short answer is that this is working as intended. The `routes` are never evaluated because `/api/user.js` is the name of your source file so the platform detected a file and served that instead of evaluating config routes.

The workaround is to change `src` to something other than the `dest` such as:

```json
{
  "routes": [
    {
      "src": "/user",
      "methods": ["POST", "GET"],
      "dest": "/api/user.js"
    }
  ]
}
```

The long answer is that zero config works differently than defining `builds`. We intentionally changed the behavior for zero config so that the file system always has precedence over user defined routes. This means the user can rename files to get different behavior if that is desired. This allows for the fewest number of routes and hopefully no configuration at all.

We also found that `methods` is kind of confusing because most real APIs return [405](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) if the method is not allowed but the endpoint exists. And that is not easy to implement with `routes`, instead the next route would be evaluated if the methods did not match.

I hope that helps!
